torch/csrc/jit/autodiff.cpp

```diff
@@ -45,22 +47,95 @@ static std::vector<Value*> gradientForNode(Node* node, ArrayRef<Value*> grad_val
+
+  std::unordered_set<Value*> requires_grad_set;
+  const auto requires_grad = [&](Value *v) { return requires_grad_set.count(v) > 0; };
+
+  auto inputs = graph.inputs();
+  for (std::size_t i = 0, num_inputs = inputs.size(); i < num_inputs; ++i) {
+    if (!input_requires_grad[i]) continue;
+    requires_grad_set.emplace(inputs[i]);
+  }
+
+  for (Node * node : graph.nodes()) {
+    if (std::none_of(node->inputs().begin(), node->inputs().end(), requires_grad)) continue;
+    for (Value * output : node->outputs())
+      requires_grad_set.emplace(output);
+  }
+
+  return requires_grad_set;
+}
+
+static Value* allocZerosLike(Value *v) {
+  static const Symbol constant_sym = "constant"_sym;
+  static const Symbol is_zero_sym  = "is_zero"_sym;
+  static const Symbol value_sym    = "value"_sym;
+  JIT_EXPECTM(v->hasType(), "can't allocate zero gradient for a value without a type");
+  Graph *graph = v->owningGraph();
+  auto type = v->type()->expect<TensorType>();
+  AutoGPU gpu_guard(type->device());
+
+  auto & at_type = type->device() == -1 ? at::CPU(type->scalarType()) : at::CUDA(type->scalarType());
+  auto zeros = at_type.zeros({1}).expand(type->sizes());
+  Node *constant = graph->create(constant_sym)
+                        ->t_(value_sym, zeros)
+                        ->i_(is_zero_sym, 1);
+  graph->appendNode(constant);
+  return constant->output();
+}
+
+struct ReverseDetails {
+  ReverseDetails(value_map&& grad_map, value_set&& requires_grad_set)
+    : grad_map(std::move(grad_map))
+    , requires_grad_set(std::move(requires_grad_set)) {}
+
+  value_map grad_map;
+  value_set requires_grad_set;
+};
+
 // Before:
 //   - graph has only stage 0
 //   - grad_desc doesn't have any fields initialized
 // After:
 //   - graph has stage 0 and stage 1 that computes its vjp
 //   - grad_desc has df_input_vjps and df_output_vjps set
 //     (but df_input_vjps will be modified later as well)
-static value_map addReverseInline(Graph& graph, Gradient& grad_desc) {
+static ReverseDetails addReverseInline(Graph& graph, Gradient& grad_desc,
+                                  const std::vector<bool>& input_requires_grad) {
   JIT_ASSERT(graph.stage() == 0);
   graph.advanceStage();
 
+  auto requires_grad_set = findAllRequiresGradNodes(graph, input_requires_grad);
+  const auto requires_grad = [&](Value *v) { return requires_grad_set.count(v) > 0; };
+
   value_map grad_map; // x -> dx mapping
-  const auto get_grad = [&](Value* v) { return grad_map.at(v); };
+  const auto get_grad = [&](Value* v) -> Value* {
+    auto it = grad_map.find(v);
+    if (it == grad_map.end()) {
+      std::tie(it, std::ignore) = grad_map.emplace(v, allocZerosLike(v));
+    }
+    return it->second;
+  };
   const auto set_grad = [&](Value *x, Value *dx) {
     if (Value * prev_grad = grad_map[x]) {
-      Value * new_grad = addAndPutAfter(prev_grad, dx, dx->node());
+      // NB: unique() gives us correct topological ordering within values created inside
```

Comment:
Relying on unique is pretty fragile to future changes to unique and to the implementation of this function. Is it every wrong to just append it to the end of the nodes list?

Reply:
hmm I guess it would be ok to just put it at the end